### PR TITLE
tests(web): implement new testing strategy phases

### PR DESCRIPTION
**Pull request by katspaugh**
> Extract, validate, and cover —
> 141 new tests guard the wall,
> fixtures match the schema,
> feature flags answer the call.

## Summary

Implements phases 1–5 of the testing strategy audit, adding ~141 new tests with zero regressions.

### Phase 1: Extract pure logic from large untested components
- **GlobalPushNotifications** (507→256 lines): 13 `_`-prefixed transform functions → `GlobalPushNotifications.utils.ts` (43 tests)
- **FirstSteps**: `calculateProgress` → `utils.ts` (6 tests)
- **ReviewStep**: 7 functions → `utils.ts` (27 tests)
- **NestedSafesPopover**: 7 functions → `utils.ts` (23 tests)

### Phase 2: Unified mock infrastructure
- `createTestServer()` factory in `config/test/msw/testServer.ts` — wraps `setupServer()` with scenario-based fixture handlers
- `renderWithScenario()` utility in `apps/web/src/tests/scenario-utils.tsx` — renders components with Redux store pre-populated from fixture scenarios, bridging Jest and Storybook mock ecosystems
- 9 example tests demonstrating all features

### Phase 3: API contract tests
- **Fixture-schema validation**: 22 tests validating all MSW fixture files against the OpenAPI schema using ajv (catches stale fixtures and backend-frontend drift)
- **Auto-generated types sync**: Hash-based check script + test ensuring `AUTO_GENERATED/` files stay in sync with `schema.json`

### Phase 4: Targeted component test coverage
- **TxNonce**: 18 tests (loading, read-only, editable, warning states)
- **FirstSteps**: 15 tests (deployed/undeployed, single/multi-sig, activation states)
- **AssetsTable**: 8 tests (empty, loading, token rendering, hidden tokens)

### Phase 5: Feature flag matrix tests
- 77 tests covering all 15 features: disabled/loading/enabled states, proxy stub behavior, handle invariants

## Test plan

- [x] `yarn workspace @safe-global/web test` — 392 suites, 3668 tests pass
- [x] `yarn workspace @safe-global/web type-check` — 0 errors
- [x] `yarn workspace @safe-global/web prettier` — clean
- [x] No regressions in existing tests

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by katspaugh** [comment]
@claude review

File: apps/web/src/features/__tests__/feature-flags-integration.test.tsx [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+jest.mock('@/hooks/useChains')
+
+// ── Feature registry ──────────────────────────────────────────────
+
+/**
+ * Each entry maps a human-readable label to its feature handle factory.
+ * Using factories (thunks) ensures the handle is created fresh per test,
+ * preventing stale module-level state from affecting results.
+ */
+const FEATURE_FLAG_MATRIX: Array<{
```
Comment (liliya-soroka) [inline apps/web/src/features/__tests__/feature-flags-integration.test.tsx:27]: I see that not all feature flags are covered . Are we going to cover only most important ? If yes, I would add a couple more

Comment (vseehausen) [reply apps/web/src/features/__tests__/feature-flags-integration.test.tsx:27]: Good point. There is a little trade off between adding all feature flags, and having to maintain the list. 

I would lean towards a list of feature flags that are necessary for the tests at hand.

So I would see the incompleteness of the feature flag list as a hint towards the incompleteness of the test coverage, and ensure to fill test coverage in the future (which will then grow the list of feature flags).

File: apps/web/src/components/balances/AssetsTable/__tests__/AssetsTable.test.tsx [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+            type: TokenType.ERC20,
+          },
+        },
+      ],
+    }
+
+    it('renders token balances in the table', async () => {
+      renderAssetsTable({ balances: twoTokenBalances })
+      await waitFor(() => {
+        expect(screen.getAllByText('100 DAI')[0]).toBeInTheDocument()
```
Comment (vseehausen) [inline apps/web/src/components/balances/AssetsTable/__tests__/AssetsTable.test.tsx]: On a general note, I am not a fan of testing for copy/text. 

I would rather test for keys here. But I guess the test is about representational functionality, so this is the exact use case.

File: apps/web/src/components/settings/PushNotifications/GlobalPushNotifications.tsx [hunk 3]
```diff
@@ -32,229 +27,45 @@ import { trackEvent } from '@/services/analytics'
...
-
-  if (safeAddresses.length !== currentNotifiedSafes[chainId].length) {
-    return false
-  }
-
-  return safeAddresses.every((safeAddress) => {
-    return currentNotifiedSafes[chainId]?.includes(safeAddress)
-  })
-}
+import {
```
Comment (vseehausen) [inline apps/web/src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: How did this refactor slip in?

File: config/test/msw/testServer.ts [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+export function createTestServer(options: TestServerOptions = {}) {
+  const { scenario, handlers: extraHandlers = [] } = options
+
+  const gatewayUrl = DEFAULT_GATEWAY_URL
+
+  const scenarioHandlers: RequestHandler[] =
+    scenario === 'empty'
+      ? [
+          ...createChainHandlersFromFixture(gatewayUrl, [FEATURES.POSITIONS, FEATURES.PORTFOLIO_ENDPOINT]),
+          ...createBalanceHandlersFromFixture('empty', gatewayUrl),
```
Comment (vseehausen) [inline config/test/msw/testServer.ts:64]: Nice, we have the MSW server here. Can't we use it in all test files? For example the balances test still creates it's own mock data.

File: apps/web/src/components/settings/PushNotifications/GlobalPushNotifications.tsx (hunk 3 above)
Comment (katspaugh) [reply apps/web/src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: Removed — reverted the entire PushNotifications refactor (utils extraction + typo fix). This should be a separate PR.

File: config/test/msw/testServer.ts (hunk 4 above)
Comment (katspaugh) [reply config/test/msw/testServer.ts:64]: Refactored the AssetsTable test to use MSW fixture data (`balancesFixtures.efSafe`, `balancesFixtures.empty`) and shared test builders (`balancesBuilder`, `balanceBuilder`, `erc20TokenBuilder`) instead of inline mock objects. The hook is still mocked via `jest.spyOn` since going through the full RTK Query pipeline would require substantial store/chain/safe setup — but the mock data now comes from shared sources.

File: apps/web/src/features/__tests__/feature-flags-integration.test.tsx (hunk 1 above)
Comment (katspaugh) [reply apps/web/src/features/__tests__/feature-flags-integration.test.tsx:27]: Agreed — the current list covers the features that have tests in this PR. As test coverage grows in future PRs, the feature flag matrix will grow with it. The incomplete list is intentional as a hint toward remaining coverage gaps, as @vseehausen noted.

File: apps/web/src/components/balances/AssetsTable/__tests__/AssetsTable.test.tsx (hunk 2 above)
Comment (katspaugh) [reply apps/web/src/components/balances/AssetsTable/__tests__/AssetsTable.test.tsx]: Acknowledged — for this test the text assertions verify the representational/rendering layer (correct headers, correct token amounts displayed). The `data-testid` assertions are used where possible (e.g. `token-balance`).

File: apps/web/src/features/__tests__/feature-flags-integration.test.tsx (hunk 1 above)
Comment (vseehausen) [reply apps/web/src/features/__tests__/feature-flags-integration.test.tsx:27]: Ok.

File: apps/web/src/components/balances/AssetsTable/__tests__/AssetsTable.test.tsx (hunk 2 above)
Comment (vseehausen) [reply apps/web/src/components/balances/AssetsTable/__tests__/AssetsTable.test.tsx]: Ok.

File: apps/web/src/components/settings/PushNotifications/GlobalPushNotifications.tsx (hunk 3 above)
Comment (vseehausen) [reply apps/web/src/components/settings/PushNotifications/GlobalPushNotifications.tsx]: Nice!

File: config/test/msw/testServer.ts (hunk 4 above)
Comment (vseehausen) [reply config/test/msw/testServer.ts:64]: Yeah, I like that!
